[RFC] vim-patch:7.4.858

fwalch opened the pull request:
Port of the ":cdo patch" from https://groups.google.com/forum/#!topic/vim_dev/dfyt-G6SMec.

Adds `:cdo` and `:ldo` commands that behave like `:argdo` and `:bufdo`, but on quickfix/location list.

Heard about this from @nelstrom on Twitter earlier today. I think these commands make a lot of sense, so I gave porting the patch to Neovim a shot. I also ported the original tests to a Lua spec.

On top of #2926, as that modifies some of the same code parts.

Original patch (excluding Makefile/.proto changes):

``` diff
diff --git a/runtime/doc/cmdline.txt b/runtime/doc/cmdline.txt
--- a/runtime/doc/cmdline.txt
+++ b/runtime/doc/cmdline.txt
@@ -511,6 +511,8 @@ followed by another Vim command:
     :argdo
     :autocmd
     :bufdo
+    :cdo
+    :cfdo
     :command
     :cscope
     :debug
@@ -521,6 +523,8 @@ followed by another Vim command:
     :help
     :helpfind
     :lcscope
+    :ldo
+    :lfdo
     :make
     :normal
     :perl
diff --git a/runtime/doc/editing.txt b/runtime/doc/editing.txt
--- a/runtime/doc/editing.txt
+++ b/runtime/doc/editing.txt
@@ -855,7 +855,8 @@ USING THE ARGUMENT LIST
            each file.
            {not in Vi} {not available when compiled without the
            |+listcmds| feature}
-           Also see |:windo|, |:tabdo| and |:bufdo|.
+           Also see |:windo|, |:tabdo|, |:bufdo|, |:cdo|, |:ldo|,
+           |:cfdo| and |:lfdo|

 Example: >
    :args *.c
diff --git a/runtime/doc/index.txt b/runtime/doc/index.txt
--- a/runtime/doc/index.txt
+++ b/runtime/doc/index.txt
@@ -1138,6 +1138,8 @@ tag         command         action ~
 |:cc|      :cc     go to specific error
 |:cclose|  :ccl[ose]   close quickfix window
 |:cd|      :cd     change directory
+|:cdo|     :cdo        execute command in each valid error list entry
+|:cfdo|        :cfdo       execute command in each file in error list
 |:center|  :ce[nter]   format lines at the center
 |:cexpr|   :cex[pr]    read errors from expr and jump to first
 |:cfile|   :cf[ile]    read file with error messages and jump to first
@@ -1296,6 +1298,8 @@ tag         command         action ~
 |:lchdir|  :lch[dir]   change directory locally
 |:lclose|  :lcl[ose]   close location window
 |:lcscope| :lcs[cope]      like ":cscope" but uses location list
+|:ldo|     :ld[o]      execute command in valid location list entries
+|:lfdo|        :lfd[o]     execute command in each file in location list
 |:left|        :le[ft]     left align lines
 |:leftabove|   :lefta[bove]    make split window appear left or above
 |:let|     :let        assign a value to a variable or option
diff --git a/runtime/doc/quickfix.txt b/runtime/doc/quickfix.txt
--- a/runtime/doc/quickfix.txt
+++ b/runtime/doc/quickfix.txt
@@ -298,6 +298,108 @@ use this code: >

    au QuickfixCmdPost make call QfMakeConv()

+EXECUTE A COMMAND IN ALL THE BUFFERS IN QUICKFIX OR LOCATION LIST:
+                           *:cdo*
+:cdo[!] {cmd}      Execute {cmd} in each valid entry in the quickfix list.
+           It works like doing this: >
+               :cfirst
+               :{cmd}
+               :cnext
+               :{cmd}
+               etc.
+<          When the current file can't be |abandon|ed and the [!]
+           is not present, the command fails.
+           When an error is detected on one buffer, further
+           buffers will not be visited.
+           The last buffer (or where an error occurred) becomes
+           the current buffer.
+           {cmd} can contain '|' to concatenate several commands.
+           Only valid entries in the quickfix list are used.
+           Note: While this command is executing, the Syntax
+           autocommand event is disabled by adding it to
+           'eventignore'.  This considerably speeds up editing
+           each buffer.
+           {not in Vi} {not available when compiled without the
+           |+listcmds| feature}
+           Also see |:bufdo|, |:tabdo|, |:argdo|, |:windo|,
+           |:ldo|, |:cfdo| and |:lfdo|.
+
+                           *:cfdo*
+:cfdo[!] {cmd}     Execute {cmd} in each file in the quickfix list.
+           It works like doing this: >
+               :cfirst
+               :{cmd}
+               :cnfile
+               :{cmd}
+               etc.
+<          When the current file can't be |abandon|ed and the [!]
+           is not present, the command fails.
+           When an error is detected on one buffer, further
+           buffers will not be visited.
+           The last buffer (or where an error occurred) becomes
+           the current buffer.
+           {cmd} can contain '|' to concatenate several commands.
+           Only valid entries in the quickfix list are used.
+           Note: While this command is executing, the Syntax
+           autocommand event is disabled by adding it to
+           'eventignore'.  This considerably speeds up editing
+           each buffer.
+           {not in Vi} {not available when compiled without the
+           |+listcmds| feature}
+           Also see |:bufdo|, |:tabdo|, |:argdo|, |:windo|,
+           |:cdo|, |:ldo| and |:lfdo|.
+
+                           *:ldo*
+:ld[o][!] {cmd}        Execute {cmd} in each entry in the location list for
+           the current window.
+           It works like doing this: >
+               :lfirst
+               :{cmd}
+               :lnext
+               :{cmd}
+               etc.
+<          When the current file can't be |abandon|ed and the [!]
+           is not present, the command fails.
+           When an error is detected on one buffer, further
+           buffers will not be visited.
+           The last buffer (or where an error occurred) becomes
+           the current buffer.
+           {cmd} can contain '|' to concatenate several commands.
+           Only valid entries in the location list are used.
+           Note: While this command is executing, the Syntax
+           autocommand event is disabled by adding it to
+           'eventignore'.  This considerably speeds up editing
+           each buffer.
+           {not in Vi} {not available when compiled without the
+           |+listcmds| feature}
+           Also see |:bufdo|, |:tabdo|, |:argdo|, |:windo|,
+           |:cdo|, |:cfdo| and |:lfdo|.
+
+                           *:lfdo*
+:lfdo[!] {cmd}     Execute {cmd} in each file in the location list for
+           the current window.
+           It works like doing this: >
+               :lfirst
+               :{cmd}
+               :lnfile
+               :{cmd}
+               etc.
+<          When the current file can't be |abandon|ed and the [!]
+           is not present, the command fails.
+           When an error is detected on one buffer, further
+           buffers will not be visited.
+           The last buffer (or where an error occurred) becomes
+           the current buffer.
+           {cmd} can contain '|' to concatenate several commands.
+           Only valid entries in the location list are used.
+           Note: While this command is executing, the Syntax
+           autocommand event is disabled by adding it to
+           'eventignore'.  This considerably speeds up editing
+           each buffer.
+           {not in Vi} {not available when compiled without the
+           |+listcmds| feature}
+           Also see |:bufdo|, |:tabdo|, |:argdo|, |:windo|,
+           |:cdo|, |:ldo| and |:cfdo|.

 =============================================================================
 2. The error window                    *quickfix-window*
diff --git a/runtime/doc/tabpage.txt b/runtime/doc/tabpage.txt
--- a/runtime/doc/tabpage.txt
+++ b/runtime/doc/tabpage.txt
@@ -248,7 +248,8 @@ LOOPING OVER TAB PAGES:
        {cmd} must not open or close tab pages or reorder them.
        {not in Vi} {not available when compiled without the
        |+listcmds| feature}
-       Also see |:windo|, |:argdo| and |:bufdo|.
+       Also see |:windo|, |:argdo|, |:bufdo|, |:cdo|, |:ldo|, |:cfdo|
+       and |:lfdo|

 ==============================================================================
 3. Other items                     *tab-page-other*
diff --git a/runtime/doc/windows.txt b/runtime/doc/windows.txt
--- a/runtime/doc/windows.txt
+++ b/runtime/doc/windows.txt
@@ -715,7 +715,8 @@ 8. Do a command in all buffers or window
            {cmd} must not open or close windows or reorder them.
            {not in Vi} {not available when compiled without the
            |+listcmds| feature}
-           Also see |:tabdo|, |:argdo| and |:bufdo|.
+           Also see |:tabdo|, |:argdo|, |:bufdo|, |:cdo|, |:ldo|,
+           |:cfdo| and |:lfdo|

                            *:bufdo*
 :[range]bufdo[!] {cmd} Execute {cmd} in each buffer in the buffer list or if
@@ -743,7 +744,8 @@ 8. Do a command in all buffers or window
            each buffer.
            {not in Vi} {not available when compiled without the
            |+listcmds| feature}
-           Also see |:tabdo|, |:argdo| and |:windo|.
+           Also see |:tabdo|, |:argdo|, |:windo|, |:cdo|, |:ldo|,
+           |:cfdo| and |:lfdo|

 Examples: >

diff --git a/src/ex_cmds.h b/src/ex_cmds.h
--- a/src/ex_cmds.h
+++ b/src/ex_cmds.h
@@ -270,12 +270,18 @@ EX(CMD_cclose,        "cclose",   ex_cclose,
 EX(CMD_cd,     "cd",       ex_cd,
            BANG|FILE1|TRLBAR|CMDWIN,
            ADDR_LINES),
+EX(CMD_cdo,        "cdo",      ex_listdo,
+           BANG|NEEDARG|EXTRA|NOTRLCOM|RANGE|NOTADR|DFLALL,
+           ADDR_LINES),
 EX(CMD_center,     "center",   ex_align,
            TRLBAR|RANGE|WHOLEFOLD|EXTRA|CMDWIN|MODIFY,
            ADDR_LINES),
 EX(CMD_cexpr,      "cexpr",    ex_cexpr,
            NEEDARG|WORD1|NOTRLCOM|TRLBAR|BANG,
            ADDR_LINES),
+EX(CMD_cfdo,       "cfdo",     ex_listdo,
+           BANG|NEEDARG|EXTRA|NOTRLCOM|RANGE|NOTADR|DFLALL,
+           ADDR_LINES),
 EX(CMD_cfile,      "cfile",    ex_cfile,
            TRLBAR|FILE1|BANG,
            ADDR_LINES),
@@ -729,6 +735,9 @@ EX(CMD_lclose,      "lclose",   ex_cclose,
 EX(CMD_lcscope,        "lcscope",  do_cscope,
            EXTRA|NOTRLCOM|XFILE,
            ADDR_LINES),
+EX(CMD_ldo,        "ldo",      ex_listdo,
+           BANG|NEEDARG|EXTRA|NOTRLCOM|RANGE|NOTADR|DFLALL,
+           ADDR_LINES),
 EX(CMD_left,       "left",     ex_align,
            TRLBAR|RANGE|WHOLEFOLD|EXTRA|CMDWIN|MODIFY,
            ADDR_LINES),
@@ -741,6 +750,9 @@ EX(CMD_let,     "let",      ex_let,
 EX(CMD_lexpr,      "lexpr",    ex_cexpr,
            NEEDARG|WORD1|NOTRLCOM|TRLBAR|BANG,
            ADDR_LINES),
+EX(CMD_lfdo,       "lfdo",     ex_listdo,
+           BANG|NEEDARG|EXTRA|NOTRLCOM|RANGE|NOTADR|DFLALL,
+           ADDR_LINES),
 EX(CMD_lfile,      "lfile",    ex_cfile,
            TRLBAR|FILE1|BANG,
            ADDR_LINES),
diff --git a/src/ex_cmds2.c b/src/ex_cmds2.c
--- a/src/ex_cmds2.c
+++ b/src/ex_cmds2.c
@@ -2429,7 +2429,7 @@ ex_argdelete(eap)
 }

 /*
- * ":argdo", ":windo", ":bufdo", ":tabdo"
+ * ":argdo", ":windo", ":bufdo", ":tabdo", ":cdo", ":ldo", ":cfdo" and ":lfdo"
  */
     void
 ex_listdo(eap)
@@ -2446,6 +2446,9 @@ ex_listdo(eap)
     char_u *save_ei = NULL;
 #endif
     char_u *p_shm_save;
+#ifdef FEAT_QUICKFIX
+    int        qf_size;
+#endif

 #ifndef FEAT_WINDOWS
     if (eap->cmdidx == CMD_windo)
@@ -2510,6 +2513,24 @@ ex_listdo(eap)
             if (buf != NULL)
        goto_buffer(eap, DOBUF_FIRST, FORWARD, buf->b_fnum);
         }
+#ifdef FEAT_QUICKFIX
+   else if (eap->cmdidx == CMD_cdo || eap->cmdidx == CMD_ldo ||
+       eap->cmdidx == CMD_cfdo || eap->cmdidx == CMD_lfdo)
+   {
+       qf_size = qf_get_size(eap);
+       if (qf_size <= 0 || eap->line1 > qf_size)
+       buf = NULL;
+       else
+       {
+       ex_cc(eap);
+
+       buf = curbuf;
+       i = eap->line1 - 1;
+       if (eap->addr_count <= 0)
+           eap->line2 = qf_size;   /* default is all the quickfix/location list entries */
+       }
+   }
+#endif
    else
        setpcmark();
    listcmd_busy = TRUE;        /* avoids setting pcmark below */
@@ -2600,6 +2621,17 @@ ex_listdo(eap)
            break;
        }

+#ifdef FEAT_QUICKFIX
+       if (eap->cmdidx == CMD_cdo || eap->cmdidx == CMD_ldo ||
+           eap->cmdidx == CMD_cfdo || eap->cmdidx == CMD_lfdo)
+       {
+       if (i >= qf_size || i >= eap->line2)
+           break;
+
+       ex_cnext(eap);
+       }
+#endif
+
        if (eap->cmdidx == CMD_windo)
        {
        validate_cursor();  /* cursor may have moved */
diff --git a/src/ex_docmd.c b/src/ex_docmd.c
--- a/src/ex_docmd.c
+++ b/src/ex_docmd.c
@@ -3838,6 +3838,8 @@ set_one_cmd_context(xp, buff)
    case CMD_botright:
    case CMD_browse:
    case CMD_bufdo:
+   case CMD_cdo:
+   case CMD_cfdo:
    case CMD_confirm:
    case CMD_debug:
    case CMD_folddoclosed:
@@ -3847,7 +3849,9 @@ set_one_cmd_context(xp, buff)
    case CMD_keepjumps:
    case CMD_keepmarks:
    case CMD_keeppatterns:
+   case CMD_ldo:
    case CMD_leftabove:
+   case CMD_lfdo:
    case CMD_lockmarks:
    case CMD_noautocmd:
    case CMD_noswapfile:
diff --git a/src/quickfix.c b/src/quickfix.c
--- a/src/quickfix.c
+++ b/src/quickfix.c
@@ -2990,19 +2990,108 @@ get_mef_name()
 }

 /*
+ * Returns the number of valid entries in the current quickfix/location list
+ */
+    int
+qf_get_size(eap)
+    exarg_T    *eap;
+{
+    qf_info_T  *qi = &ql_info;
+    qfline_T   *qfp;
+    int        i, sz = 0;
+    int        prev_fnum = 0;
+
+    if (eap->cmdidx == CMD_ldo || eap->cmdidx == CMD_lfdo)
+    {
+   /* Location list */
+   qi = GET_LOC_LIST(curwin);
+   if (qi == NULL)
+       return 0;
+    }
+
+    for (i = 0, qfp = qi->qf_lists[qi->qf_curlist].qf_start;
+       (i < qi->qf_lists[qi->qf_curlist].qf_count) && (qfp != NULL);
+       ++i, qfp = qfp->qf_next)
+    {
+       if (qfp->qf_valid)
+       if (eap->cmdidx == CMD_cdo || eap->cmdidx == CMD_ldo)
+           sz++;   /* Count all valid entries */
+       else if (qfp->qf_fnum > 0 && qfp->qf_fnum != prev_fnum)
+       {
+           /* Count the number of files */
+           sz++;
+           prev_fnum = qfp->qf_fnum;
+       }
+    }
+
+    return sz;
+}
+
+/*
+ * Get the 'n'th valid error entry in the quickfix or location list.
+ * Used by :cdo, :ldo, :cfdo and :lfdo commands.
+ * For :cdo and :ldo returns the 'n'th valid error entry.
+ * For :cfdo and :lfdo returns the 'n'th valid file entry.
+ */
+    static int
+qf_get_nth_valid_entry(qi, n, fdo)
+    qf_info_T  *qi;
+    int        n;
+    int        fdo;
+{
+    qf_list_T  *qfl = &qi->qf_lists[qi->qf_curlist];
+    qfline_T   *qfp = qfl->qf_start;
+    int        i, eidx;
+    int        prev_fnum = 0;
+
+    /* check if the list has valid errors */
+    if (qfl->qf_count <= 0 || qfl->qf_nonevalid)
+   return 1;
+
+    for (i = 1, eidx = 0; i <= qfl->qf_count && qfp!= NULL;
+       i++, qfp = qfp->qf_next)
+    {
+   if (qfp->qf_valid)
+       if (fdo)
+       {
+       if (qfp->qf_fnum > 0 && qfp->qf_fnum != prev_fnum)
+       {
+           /* Count the number of files */
+           eidx++;
+           prev_fnum = qfp->qf_fnum;
+       }
+       }
+       else
+       eidx++;
+
+   if (eidx == n)
+       break;
+    }
+
+    if (i <= qfl->qf_count)
+   return i;
+    else
+   return 1;
+}
+
+/*
  * ":cc", ":crewind", ":cfirst" and ":clast".
  * ":ll", ":lrewind", ":lfirst" and ":llast".
+ * ":cdo", ":ldo", ":cfdo" and ":lfdo"
  */
     void
 ex_cc(eap)
     exarg_T    *eap;
 {
     qf_info_T  *qi = &ql_info;
+    int        errornr;

     if (eap->cmdidx == CMD_ll
        || eap->cmdidx == CMD_lrewind
        || eap->cmdidx == CMD_lfirst
-       || eap->cmdidx == CMD_llast)
+       || eap->cmdidx == CMD_llast
+       || eap->cmdidx == CMD_ldo
+       || eap->cmdidx == CMD_lfdo)
     {
    qi = GET_LOC_LIST(curwin);
    if (qi == NULL)
@@ -3012,34 +3101,51 @@ ex_cc(eap)
    }
     }

-    qf_jump(qi, 0,
-       eap->addr_count > 0
-       ? (int)eap->line2
-       : (eap->cmdidx == CMD_cc || eap->cmdidx == CMD_ll)
-       ? 0
-       : (eap->cmdidx == CMD_crewind || eap->cmdidx == CMD_lrewind
-          || eap->cmdidx == CMD_cfirst || eap->cmdidx == CMD_lfirst)
-           ? 1
-           : 32767,
-       eap->forceit);
+    if (eap->addr_count > 0)
+   errornr = (int)eap->line2;
+    else
+    {
+   if (eap->cmdidx == CMD_cc || eap->cmdidx == CMD_ll)
+       errornr = 0;
+   else if (eap->cmdidx == CMD_crewind || eap->cmdidx == CMD_lrewind
+       || eap->cmdidx == CMD_cfirst || eap->cmdidx == CMD_lfirst)
+       errornr = 1;
+   else
+       errornr = 32767;
+    }
+
+    /* For cdo and ldo commands, jump to the nth valid error.
+     * For cfdo and lfdo commands, jump to the nth valid file entry.
+     */
+    if (eap->cmdidx == CMD_cdo || eap->cmdidx == CMD_ldo ||
+       eap->cmdidx == CMD_cfdo || eap->cmdidx == CMD_lfdo)
+   errornr = qf_get_nth_valid_entry(qi,
+       eap->addr_count > 0 ? (int)eap->line1 : 1,
+       eap->cmdidx == CMD_cfdo || eap->cmdidx == CMD_lfdo);
+
+    qf_jump(qi, 0, errornr, eap->forceit);
 }

 /*
  * ":cnext", ":cnfile", ":cNext" and ":cprevious".
  * ":lnext", ":lNext", ":lprevious", ":lnfile", ":lNfile" and ":lpfile".
+ * Also, used by ":cdo", ":ldo", ":cfdo" and ":lfdo" commands.
  */
     void
 ex_cnext(eap)
     exarg_T    *eap;
 {
     qf_info_T  *qi = &ql_info;
+    int        errornr;

     if (eap->cmdidx == CMD_lnext
        || eap->cmdidx == CMD_lNext
        || eap->cmdidx == CMD_lprevious
        || eap->cmdidx == CMD_lnfile
        || eap->cmdidx == CMD_lNfile
-       || eap->cmdidx == CMD_lpfile)
+       || eap->cmdidx == CMD_lpfile
+       || eap->cmdidx == CMD_ldo
+       || eap->cmdidx == CMD_lfdo)
     {
    qi = GET_LOC_LIST(curwin);
    if (qi == NULL)
@@ -3049,15 +3155,24 @@ ex_cnext(eap)
    }
     }

-    qf_jump(qi, (eap->cmdidx == CMD_cnext || eap->cmdidx == CMD_lnext)
+    if (eap->addr_count > 0 &&
+       (eap->cmdidx != CMD_cdo && eap->cmdidx != CMD_ldo &&
+        eap->cmdidx != CMD_cfdo && eap->cmdidx != CMD_lfdo))
+   errornr = (int)eap->line2;
+    else
+   errornr = 1;
+
+    qf_jump(qi, (eap->cmdidx == CMD_cnext || eap->cmdidx == CMD_lnext
+       || eap->cmdidx == CMD_cdo || eap->cmdidx == CMD_ldo)
        ? FORWARD
-       : (eap->cmdidx == CMD_cnfile || eap->cmdidx == CMD_lnfile)
+       : (eap->cmdidx == CMD_cnfile || eap->cmdidx == CMD_lnfile
+       || eap->cmdidx == CMD_cfdo || eap->cmdidx == CMD_lfdo)
        ? FORWARD_FILE
        : (eap->cmdidx == CMD_cpfile || eap->cmdidx == CMD_lpfile
           || eap->cmdidx == CMD_cNfile || eap->cmdidx == CMD_lNfile)
            ? BACKWARD_FILE
            : BACKWARD,
-       eap->addr_count > 0 ? (int)eap->line2 : 1, eap->forceit);
+       errornr, eap->forceit);
 }

 /*
```

nelstrom commented:
I'm glad to see this!

elmart commented:
A huge +1 on this!

fwalch commented:
test10 failed, but it's also a problem in the original patch: https://travis-ci.org/fwalch/vim-ci/builds/68879635

fwalch commented:
As suggested by @chrisbra (thanks!) on `vim_dev`, swapped the order of the commands in `ex_cmds.lua`. This fixes the `test10` error.

justinmk commented:
@fwalch The author mentioned on vim_dev that he would soon post an update with support for ranges (or something like that). Shall we wait for that?

fwalch commented:
Updated for new version, according to [interdiff](https://gist.github.com/fwalch/4853ddb4505aafb7b5c1).

pwntester commented:
Seems like it has already been integrated into vim:
https://twitter.com/nelstrom/status/642262600676450304

justinmk commented:
Yep, this needs an update I assume.

fwalch commented:
The test hangs because of a "Press ENTER ..." prompt. Swapping the `execute("call RunTests('l')")` with `feed(":call RunTests('l')<cr><cr>")` removes the hanging, but causes very strange test results. The legacy test (`make test_cdo.out`; not included here) works. Would be glad for any hints..

justinmk commented:
Try `<c-l>` instead of the two CRs.

fwalch commented:
@justinmk That works for getting rid of the hanging, thanks! Now to find out why the buffer contents are wrong.. works if I try the test steps manually.

Shougo commented:
@fwalch Why the PR does not change version.c?

fwalch commented:
@Shougo Thanks, I will fix that. It's because this PR was originally created when the patch was not in Vim.

Shougo commented:
Oh, thanks.  I get it.

Tranquility commented:
Don't know why the tests failed but I would love to see this one merged.

justinmk commented:
Me too. @fwalch can you rebase?

fwalch commented:
Sure, rebased. Unfortunately, I still couldn't figure out why the test fails.

fwalch committed:
vim-patch:7.4.858

Problem:    It's a bit clumsy to execute a command on a list of matches.
Solution:   Add the ":ldo", ":lfdo", ":cdo" and ":cfdo" commands. (Yegappan
            Lakshmanan)

https://github.com/vim/vim/commit/aa23b379421aa214e6543b06c974594a25799b09

justinmk commented on test/functional/legacy/cdo_spec.lua:
```diff
@@ -0,0 +1,198 @@
+  it('works for :cdo', function()
+    execute("let result=''")
+
+    feed(":call RunTests('c')<c-l>")
```
This also avoids the hang:

    execute([[call RunTests('c')]])
    helpers.run(nil, nil, nil, 100)

justinmk commented on test/functional/legacy/cdo_spec.lua:
```diff
@@ -0,0 +1,198 @@
+    feed(":call RunTests('c')<c-l>")
+
+    execute("edit! cdo.out")
+    execute("0put =result")
```
These two lines are being ignored. If one replaces them with just `execute("echo ''")`, the result is the same.

justinmk commented:
I found [another way](https://github.com/neovim/neovim/pull/2927#discussion_r46790505) to avoid the hang, but it yields the same wrong results:

```
xLine1
Line2
Line3
```

Things get more strange:

- Adding a call like `print('g:result='..helpers.eval('g:result'))` again causes a hang
- The final two calls in the test aren't running at all. 

I don't see anything suspicious in this PR, but if things were this unreliable on master I can't see how the other tests would be passing...

kassio commented:
Any idea when will this get in? I miss this feature soo much! This is the only reason sometimes I open `vim` instead of `neovim`.

BohrShaw commented:
> This is the only reason sometimes I open vim instead of neovim.

So true.

justinmk commented:
Yes, I'd like this patch too. I [looked into the failing test](https://github.com/neovim/neovim/pull/2927#issuecomment-162439263) for hours. Once I or someone else figures it out we can merge it...

fwalch commented:
I noticed that the problems occur only when using `--embed`, not `--headless`. Applying the following patch seems to "fix" the test hang for me.. let's see if Travis agrees. And then we only have to figure out how to fix the problem for real... :wink:

```diff
diff --git a/src/nvim/message.c b/src/nvim/message.c
index 66b8b9b..5f26bfa 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -1964,7 +1964,7 @@ static void t_puts(int *t_col, char_u *t_s, char_u *s, int attr)
 // happens when no UIs are attached and nvim is not being embedded
 int msg_use_printf(void)
 {
-  return !embedded_mode && !ui_active();
+  return !ui_active();
 }
 
 /*
```

justinmk commented:
@fwalch That's a good hint!

fwalch commented:
Let's take the easy way out for now to move this forward. I removed the Lua tests and added the original legacy tests. I'll create a separate issue to further investigate the problems with `--embed` resp. why the ported tests are failing.

Tranquility commented:
:+1:

justinmk commented:
Verified the tests with 

    make clean && TESTNUM=_cdo make

after updating `testdir/Makefile`.

kassio commented:
❤️💛💙💚💜
